cosmpy/aerial/client/utils.py

```diff
@@ -18,11 +18,106 @@
 # ------------------------------------------------------------------------------
 """Helper functions."""
 from datetime import timedelta
-from typing import Any, Callable, List, Optional, Union
+from typing import Any, Callable, List, Optional, Tuple, Union
 
-from cosmpy.aerial.tx import SigningCfg
+from cosmpy.aerial.coins import parse_coins
+from cosmpy.aerial.tx import SigningCfg, Transaction
 from cosmpy.aerial.tx_helpers import SubmittedTx
+from cosmpy.crypto.address import Address
 from cosmpy.protos.cosmos.base.query.v1beta1.pagination_pb2 import PageRequest
+from cosmpy.protos.cosmos.tx.v1beta1.tx_pb2 import Fee
+
+
+def simulate_tx(
+    client: "LedgerClient",  # type: ignore # noqa: F821
+    tx: Transaction,
+    sender: "Wallet",  # type: ignore # noqa: F821
+    account: "Account",  # type: ignore # noqa: F821
+    memo: Optional[str] = None,
+) -> Tuple[int, str]:
+    """Estimate transaction fees based on either a provided amount, gas limit, or simulation.
+
+    :param client: Ledger client
+    :param tx: The transaction
+    :param sender: The transaction sender
+    :param account: The account
+    :param memo: Transaction memo, defaults to None
+
+    :return: Estimated gas_limit and fee amount tuple
+    """
+    # we need to build up a representative transaction so that we can accurately simulate it
+    tx.seal(
+        SigningCfg.direct(sender.public_key(), account.sequence),
+        fee="",
+        gas_limit=0,
+        memo=memo,
+    )
+    tx.sign(sender.signer(), client.network_config.chain_id, account.number)
+    tx.complete()
+
+    # simulate the gas and fee for the transaction
+    gas_limit, fee = client.estimate_gas_and_fee_for_tx(tx)
+
+    return gas_limit, fee
+
+
+def estimate_tx_fees(
+    client: "LedgerClient",  # type: ignore # noqa: F821
+    tx: Transaction,
+    sender: "Wallet",  # type: ignore # noqa: F821
+    amount: Optional[str] = None,
+    gas_limit: Optional[int] = None,
+    granter: Optional[Address] = None,
+    account: Optional["Account"] = None,  # type: ignore # noqa: F821
+    memo: Optional[str] = None,
+) -> Tuple[Fee, Optional["Account"]]:  # type: ignore # noqa: F821
+    """Estimate transaction fees based on either a provided amount, gas limit, or simulation.
+
+    :param client: Ledger client
+    :param tx: The transaction
+    :param sender: The transaction sender
+    :param amount: Transaction fee amount, defaults to None
+    :param gas_limit: The gas limit
+    :param granter: Transaction fee granter, defaults to None
+    :param account: The account
+    :param memo: Transaction memo, defaults to None
+
+    :return: Fee object and queried account tuple
+    """
+    # CASE 1: Fee amount is already provided
+    if amount:
+        return (
+            Fee(
+                amount=parse_coins(amount),
+                gas_limit=gas_limit,
+                granter=granter,
```

Comment:
granter can be None

Reply:
I will have to double check this, it was None in previous implementation.

Reply:
Granter is the external wallet with funds, so `payer` can be None. Payer is named`grantee` in newer verision of CosmosSDK.